### ISSUE TITLE
Automatically adjust the number of columns to screen size

### ISSUE THREAD

**Issue by xi7** [falzonv/discreet-launcher#67]
![Screenshot_2021-06-09-07-28-51](https://user-images.githubusercontent.com/67331785/121339270-af169400-c8f4-11eb-8e64-fb5177c6273a.png)

https://user-images.githubusercontent.com/67331785/121337697-3105bd80-c8f3-11eb-99a1-1db7f0500976.mp4

row landscape = 7 or 8 and not 5 (and also folder with fit windows)

**or possibility change by settings**

**Comment by xi7** [comment]
and when possible test the home favorites only with icons but fit layout size.

like this ex: icons = 100 , row 6, folder have name.

**always consider better screen use**

**Comment by xi7** [comment]
![Screenshot_2021-06-09-08-48-24](https://user-images.githubusercontent.com/67331785/121349511-fc4c3300-c8ff-11eb-8335-31c28f9e0680.png)
portrait mode - home favorites
waste of layout, half of screen with almost nothing
**this is a favorites selected manually, i think no need be a mirror of grid_view with only 4 row and names and small icon.**

**Comment by xi7** [comment]
![Screenshot_2021-06-10-07-34-09](https://user-images.githubusercontent.com/67331785/121517255-85796d80-c9c5-11eb-8675-0e65516c26c3.png)

portrait w/ 7 and no names.

**Comment by falzonv** [comment]
Hello,

Ok, thank you for the screenshots, videos and explanations :-)

The problem with a fixed number of columns is that it would not work well in all cases, for example 7 would be too much for some small screens, and I agree that 5 is too low on bigger screen.

I will try to have the number of columns automatically adjusted to the screen width (and updated when the user rotates the screen), and also updated when the option to hide app names is enabled/disabled.

This will probably be pat of the next release.

---

For now I like the current favorites panel and do not plan to change it. Indeed it may take a bit of space depending on the number of icons, but it is hidden by default which leaves the home screen completely empty.

Also, the notification popup already provides a more condensed view of the favorites.
*To be honest this was not intended but, because it is displayed over other applications, I have very few control over the appearance of the notification popup (I cannot even set it full width to match the usual folders: Android automatically adds margins around the popup when it is displayed).*

Best regards.

**Comment by xi7** [comment]
my phone have 4 inch,  work very well with 10x landscape and 6x portrait, agradable visualization and usable.

**Comment by xi7** [comment]
the columns to screen size is only while no names. 
with names fix only the landscape with 7x and portrait 4x.  
if no work automatically enable function manually.

*possible fix the notification area size folder view ? use the same size grid_view folder.

**Comment by xi7** [comment]
possible make 1 gesture and call the folder notification ?

**Comment by falzonv** [comment]
> the columns to screen size is only while no names.
with names fix only the landscape with 7x and portrait 4x.
if no work automatically enable function manually.

I am bit surprised because for other things you were very pushy to have me remove/reduce code already optimized, and here I work on a solution which should cover all cases simply (portrait/landscape, with/without names) and you would prefer to have two different behaviors (with/without names), perhaps with an additional setting, etc. That is not very logical.

The first small tests are quite positive anyway, so I plan to do a common solution based on screen size.

In addition, I am thinking about adding an option "Remove margins". Personally I don't like the applications being too close like on your last screenshot (maybe I have big fingers :joy:), but you seem to prefer it (to each his own).
=> Would you like to have an option like this?

---

> *possible fix the notification area size folder view ? use the same size grid_view folder.

For now I was not able to fix this unfortunately, this is also something I am not really happy with :-/
This is what I explained in a previous comment:

> because it is displayed over other applications, I have very few control over the appearance of the notification popup (I cannot even set it full width to match the usual folders: Android automatically adds margins around the popup when it is displayed).

I have made many tries but for now there were only two results:
- The current notification with the added margins over the previous application
- Or a notification exactly like the folder, but the background is black and it acts like a come back to home screen

The best would be a mix of the two, but for now I was not able to do it (tried a couple of hours, but there are also many other requests so I decided to let it like this for now).

---

> possible make 1 gesture and call the folder notification ?

From the home screen: yes, this is already existing with the favorites panel.
From another application: no, I am not able to do this sorry.

**Comment by falzonv** [comment]
After further tests, I think the result will be something like this:
- Number of columns adapted to screen size
  - Should give approximately 4x portrait and 7x landscape with most phones (more with large phones)
  - No difference with/without apps names, because I realized some users may still prefer spaced even when no names
- Additional option to remove margins, accessible only when the user has selected "no names" for the apps
  - This should give a result similar to your last screenshot

At the end, it is not very far from what you suggested :-)

**Comment by xi7** [comment]
> I am bit surprised because for other things you were very pushy to have me remove/reduce code already optimized, and here I work on a solution which should cover all cases simply (portrait/landscape, with/without names) and you would prefer to have two different behaviors (with/without names), perhaps with an additional setting, etc. That is not very logical.

i just change smali file 0x4 to 0x6 portrait and 0x5 to 0x10 landscape. (no names situation)

**Comment by xi7** [comment]
have print notification test ?

**Comment by xi7** [comment]
https://user-images.githubusercontent.com/67331785/121743632-058ff800-cad8-11eb-8ad6-36a4c6005d84.mp4

that my problem with notification area, very small cant scroll icons

**Comment by xi7** [comment]
https://user-images.githubusercontent.com/67331785/121771364-50942480-cb45-11eb-8609-edba6f7a2f36.mp4
landscape x9 icons size 60dp

https://user-images.githubusercontent.com/67331785/121771394-7de0d280-cb45-11eb-9820-32068e6019c6.mp4
portrait x5 icons size 60dp

-------
**#67 no names good solution, big and organized

about with names i dont have solution because the system interferes with the font size**

**Comment by xi7** [comment]
![Screenshot_2021-06-13-09-57-32](https://user-images.githubusercontent.com/67331785/121808455-f8395180-cc2e-11eb-954d-78431d41082a.png)

manually full solution

**Comment by xi7** [comment]
![Screenshot_2021-06-13-09-57-32](https://user-images.githubusercontent.com/67331785/121808616-b361ea80-cc2f-11eb-9714-d97bc1d01725.png)

manually full solution

**Comment by falzonv** [comment]
> possible fix the notification area size folder view ? use the same size grid_view folder.

=> Good news, after comparing dozens of parameters in Android styles and many more tests, the notification width is now fixed!
It seems to work well on all versions, but on my phone I had to perform a reboot before it was ok, strange...
Anyway, the issue with the width will be solved in the next release :-)

> that my problem with notification area, very small cant scroll icons

=> Ok, I will set a minimum height to the notification popup and it should solve this issue.

\
Regarding the last screenshot, sorry but I do not plan to implement all these options (except hide apps name which is already done, and I will add an option to remove or not margins when apps names are hidden).
Most users do not need such detail of customization (text size, icon size, ...) to be happy with the launcher and it would add a lot of complexity (and possible bugs) in the code for future maintenance.

**Comment by xi7** [comment]
> Na tela inicial: sim, isso já existe com o painel de favoritos.
> De outro aplicativo: não, não posso fazer isso, desculpe.

is not possible the button home open the notification folder without permission when out home ?

**Comment by xi7** [comment]
> > the columns to screen size is only while no names.
> > with names fix only the landscape with 7x and portrait 4x.
> > if no work automatically enable function manually.
> 
> I am bit surprised because for other things you were very pushy to have me remove/reduce code already optimized, and here I work on a solution which should cover all cases simply (portrait/landscape, with/without names) and you would prefer to have two different behaviors (with/without names), perhaps with an additional setting, etc. That is not very logical.
> 
> The first small tests are quite positive anyway, so I plan to do a common solution based on screen size.
> 
> In addition, I am thinking about adding an option "Remove margins". Personally I don't like the applications being too close like on your last screenshot (maybe I have big fingers 😂), but you seem to prefer it (to each his own).
> => Would you like to have an option like this?
> 
> > *possible fix the notification area size folder view ? use the same size grid_view folder.
> 
> For now I was not able to fix this unfortunately, this is also something I am not really happy with :-/
> This is what I explained in a previous comment:
> 
> > because it is displayed over other applications, I have very few control over the appearance of the notification popup (I cannot even set it full width to match the usual folders: Android automatically adds margins around the popup when it is displayed).
> 
> I have made many tries but for now there were only two results:
> 
> * The current notification with the added margins over the previous application
> * Or a notification exactly like the folder, but the background is black and it acts like a come back to home screen
> 
> The best would be a mix of the two, but for now I was not able to do it (tried a couple of hours, but there are also many other requests so I decided to let it like this for now).
> 
> > possible make 1 gesture and call the folder notification ?
> 
> From the home screen: yes, this is already existing with the favorites panel.
> From another application: no, I am not able to do this sorry.

**Comment by xi7** [comment]
sorry i clicked wrong and close, how to back ?

**Comment by falzonv** [comment]
> is not possible the button home open the notification folder without permission when out home ?

No sorry, this is not possible to override the home button.
I think this is to prevent malicious apps from blocking the user phone (if it overrides the home button, the back button and maybe some others, the user will not be able to leave the malicious app without rebooting the phone ; and if the app has permission to run on startup, then the user is really locked).

---

Regarding other things from this ticket, I am pleased to inform you that this has been implemented in new release v4.1.0!
This release has been published a few minutes ago.

Here are the usual availability delays for each location:
- 4 to 8 days on F-Droid
- 0 to 2 days on Google Play
- Already available as direct APK on GitHub (https://github.com/falzonv/discreet-launcher/releases)

*Remarks:
I will leave this ticket opened until the release is effectively published on F-Droid.
You can use the export/import feature to switch from one version to another.*

Best regards.

**Comment by xi7** [comment]
4.1.0 perfect !

**Comment by xi7** [comment]
![Screenshot_2021-06-19-07-42-58](https://user-images.githubusercontent.com/67331785/122640337-72595280-d0d5-11eb-9756-866d26da9b47.png)

android 5 notification area i have target black

**Comment by falzonv** [comment]
Thank you for your feedback! :-)

> android 5 notification area i have target black folder

Strange, I was not yet able to reproduce it with the Android 5 emulator :-/
(I have a small shadow added by Android above and below the notification area, but not black rectangles like on your screenshot)

Do you also observe this bug when opening a folder from the notification area?

**Comment by xi7** [comment]
https://user-images.githubusercontent.com/67331785/122644385-4ea10700-d0eb-11eb-9971-5576030f1697.mp4

here show only black target

**Comment by xi7** [comment]
home favorites work perfect but with notifications area black and a big white text area while landscape

**Comment by falzonv** [comment]
> android 5 notification area i have target black folder

=> I tried something to fix this in v4.1.1 if you want to check.
On my side it removed the shadow, so I hope it will remove the black for you.
If not, I have no other idea for now :-/

> big white text area while landscape

=> For me this happens only when clicking the search icon in landscape mode, and only with some keyboards (for example, it does it with the standard Google keyboard but not with [Hacker's Keyboard](https://f-droid.org/en/packages/org.pocketworkstation.pckeyboard/)). This is not a bug I will be able to fix for the search feature, as it is not on Discreet Launcher side but on keyboard side.

The keyboard should be shown only after you clicked the search icon, not immediately when opening the notification.

**Comment by xi7** [comment]
app package problem, i cant install 4.1.1

**Comment by falzonv** [comment]
Ok, thank you for the report!
After regenerating the APK, the size is slightly different so probably something was missing (but no idea what, I clicked the same generation button both times and there was no changes in between...).

Anyway, I replaced the APK file from the Releases page and now it should work normally.

**Comment by xi7** [comment]
problem solved now. working.

**Comment by xi7** [comment]
> => For me this happens only when clicking the search icon in landscape mode, and only with some keyboards (for example, it does it with the standard Google keyboard but not with [Hacker's Keyboard](https://f-droid.org/en/packages/org.pocketworkstation.pckeyboard/)). This is not a bug I will be able to fix for the search feature, as it is not on Discreet Launcher side but on keyboard side.
> 
> The keyboard should be shown only after you clicked the search icon, not immediately when opening the notification.

after test with a lot keyboard, this is a layout problem you need fix it. (landscape mode)
the portrait work fine.

**Comment by xi7** [comment]
https://user-images.githubusercontent.com/67331785/122759654-9309df80-d270-11eb-87c5-cde115dcc83f.mp4
landscape bad layout

https://user-images.githubusercontent.com/67331785/122759692-9bfab100-d270-11eb-80f3-9850bce59c53.mp4
portrait

**Comment by falzonv** [comment]
Ok, thank you for the tests and feedback.
I have found a workaround that should work with most keyboards, it will be implemented in the next release.

**Comment by xi7** [comment]
maybe always portrait search solve.

**Comment by xi7** [comment]
its possible choose manually what to do the swipe up/down/ and notification ?
-direct search
-show home favorites
-show grid view
-and notification... favorites or direct search

**Comment by falzonv** [comment]
Sorry, the up/down navigation is not something that I plan to change.

If I do ticket #73, it would add possibility to launch one application with "swipe left" or "swipe right", so the user could assign the "Search" application to one of these swipes and have direct search from the home screen in only one swipe.

Regarding the notification opening favorites or direct search, I will have to check.
If this can be done without too much duplicate code I am ok to implement this option.
=> Created ticket #75 for follow-up.

---

Since Thursday, release v4.1.1 is available on F-Droid so I will close this ticket.
Feel free to add comments if needed or wished.
